### PR TITLE
ARROW-18157: [Dev][Archery] "archery docker run" sets env var to None when inherited

### PR DESCRIPTION
**Pull request by pitrou**
This would occur when:
* an environment variable is set to inherited in `docker-compose.yml`
* the variable isn't set in the calling environment
* Archery uses `docker` instead of `docker-compose` to run the container (for example because it's a GPU job)

### CODE REVIEW

**Comment by pitrou** [comment]
cc @assignUser @raulcd

**Comment by raulcd** [comment]
Shouldn't we just fail if the environment variable was not found? It feels like a problem on the configuration as if the env var is not needed why was there on the first place?

**Comment by assignUser** [comment]
No, the behavior of setting it to `None` is contra to how docker-compose handles things. Having a key with an empty value means "inherit this from the calling env,  if unset then ignore it". That should happen with archery in the same way.

File: dev/archery/archery/docker/core.py [hunk 1]
```diff
@@ -342,7 +342,8 @@ def run(self, service_name, command=None, *, env=None, volumes=None,
 
             # append env variables from the compose conf
             for k, v in service.get('environment', {}).items():
-                args.extend(['-e', '{}={}'.format(k, v)])
+                if v is not None:
+                    args.extend(['-e', '{}={}'.format(k, v)])
```
Comment (kou) [inline dev/archery/archery/docker/core.py:346]: It seems that we need to get an environment variable value when `v` is `None` to emulate the `docker-compose` behavior:

```suggestion
                if v is None:
                  v = os.environ.get(k)
                if v is not None:
                    args.extend(['-e', '{}={}'.format(k, v)])
```

Comment (pitrou) [reply dev/archery/archery/docker/core.py:346]: Are you sure? It didn't seem needed when I tried locally.

Comment (pitrou) [reply dev/archery/archery/docker/core.py:346]: I tested again locally and it works (using `SETUPTOOLS_SCM_PRETEND_VERSION=1.2.3 archery --debug  docker  run ubuntu-cuda-python bash`). Do you get a different result?
```console
$ SETUPTOOLS_SCM_PRETEND_VERSION=1.2.3 archery --debug  docker  run ubuntu-cuda-python bash
DEBUG:archery:Executing `['git', '-C', '/home/antoine/arrow/dev/cpp', 'rev-parse', '--show-toplevel']`
DEBUG:archery:Executing `['docker-compose', '--file', '/home/antoine/arrow/dev/docker-compose.yml', 'config']`
DEBUG:archery:Executing `['docker-compose', '--file', '/home/antoine/arrow/dev/docker-compose.yml', 'pull', '--ignore-pull-failures', 'ubuntu-cuda-cpp']`
Pulling ubuntu-cuda-cpp ... done
DEBUG:archery:Executing `['docker-compose', '--file', '/home/antoine/arrow/dev/docker-compose.yml', 'pull', '--ignore-pull-failures', 'ubuntu-cuda-python']`
Pulling ubuntu-cuda-python ... done
DEBUG:archery:Executing `['docker-compose', '--file', '/home/antoine/arrow/dev/docker-compose.yml', 'build', '--build-arg', 'BUILDKIT_INLINE_CACHE=1', 'ubuntu-cuda-cpp']`
...
DEBUG:archery:Executing `['docker-compose', '--file', '/home/antoine/arrow/dev/docker-compose.yml', 'build', '--build-arg', 'BUILDKIT_INLINE_CACHE=1', 'ubuntu-cuda-python']`
...
DEBUG:archery:Executing `['docker', 'run', '--rm', '--gpus', 'all', '--shm-size', '2G', '-e', 'ARROW_BUILD_UTILITIES=OFF', '-e', 'ARROW_COMPUTE=ON', '-e', 'ARROW_CSV=ON', '-e', 'ARROW_CUDA=ON', '-e', 'ARROW_DATASET=ON', '-e', 'ARROW_ENABLE_TIMING_TESTS=OFF', '-e', 'ARROW_FILESYSTEM=ON', '-e', 'ARROW_GANDIVA=OFF', '-e', 'ARROW_GCS=OFF', '-e', 'ARROW_HDFS=ON', '-e', 'ARROW_JEMALLOC=ON', '-e', 'ARROW_JSON=ON', '-e', 'ARROW_ORC=OFF', '-e', 'ARROW_PARQUET=ON', '-e', 'ARROW_PLASMA=OFF', '-e', 'ARROW_S3=OFF', '-e', 'ARROW_SUBSTRAIT=OFF', '-e', 'ARROW_WITH_OPENTELEMETRY=OFF', '-e', 'CCACHE_COMPILERCHECK=content', '-e', 'CCACHE_COMPRESS=1', '-e', 'CCACHE_COMPRESSLEVEL=6', '-e', 'CCACHE_DIR=/ccache', '-e', 'CCACHE_MAXSIZE=1G', '-e', 'SCCACHE_S3_KEY_PREFIX=sccache', '-e', 'SETUPTOOLS_SCM_PRETEND_VERSION=1.2.3', '-v', '/home/antoine/arrow/dev:/arrow:delegated', '-v', 'ubuntu-ccache:/ccache:delegated', '-it', 'apache/arrow-dev:amd64-ubuntu-20.04-cuda-11.0.3-python-3', 'bash']`
root@6e458d15abf9:/# echo $SETUPTOOLS_SCM_PRETEND_VERSION 
1.2.3

```

Comment (assignUser) [reply dev/archery/archery/docker/core.py:346]: archery already gets the enclosing env vars: 
https://github.com/apache/arrow/blob/master/dev/archery/archery/docker/core.py#L81 
so no need to do it manually here.

**Review by assignUser — APPROVED** [review]
+1

File: dev/archery/archery/docker/core.py (hunk 1 above)
Comment (kou) [reply dev/archery/archery/docker/core.py:346]: Ah, sorry. I just saw only code around this.

**Pull request merged**